### DISCUSSION TITLE
Adding first letter uppercase selector support (::first-letter)

### DISCUSSION THREAD

**Discussion by damikun** [tailwindlabs/tailwindcss#2439] [Ideas]
It can be nice to have build-in "first letter uppercase class", under text-transform..

Provide ability to  use _CSS selector_  ` :first-letter` 

https://www.w3schools.com/cssref/sel_firstletter.asp

**Comment by neupauer** [comment]
Hi, I think, you can achieve that, by adding [custom variants](https://tailwindcss.com/docs/pseudo-class-variants#creating-custom-variants)

And then you can use `first-letter:whatever ` e.g., `first-letter:uppercase `

**Comment by simonswiss** [comment]
Hey @damikun 👋

As mentioned above, this is a good use case for a plugin that adds a `first-letter` variant!

Here's an example to define the plugin:

```javascript
const plugin = require('tailwindcss/plugin')

module.exports = {
  purge: [],
  theme: {},
  variants: {},
  plugins: [
    plugin(function ({ addVariant, e }) {
      addVariant('first-letter', ({ modifySelectors, separator }) => {
        modifySelectors(({ className }) => {
          return `.${e(`first-letter${separator}${className}`)}:first-letter`
        })
      })
    }),
  ],
}
```

This alone won't do anything, as you would need to enable the variant on specific core plugins.

Here, I will add it to the `textTransform` plugin:

```javascript
variants: {
  textTransform: ({ after }) => after(['first-letter']),
},
```

Together, here's the config file:

```javascript
const plugin = require('tailwindcss/plugin')

module.exports = {
  purge: [],
  theme: {},
  variants: {
    textTransform: ({ after }) => after(['first-letter']),
  },
  plugins: [
    plugin(function ({ addVariant, e }) {
      addVariant('first-letter', ({ modifySelectors, separator }) => {
        modifySelectors(({ className }) => {
          return `.${e(`first-letter${separator}${className}`)}:first-letter`
        })
      })
    }),
  ],
}
```

With this, you are now able to `uppercase` the first letter of a paragraph for example, by prefixing the `uppercase` with `first-letter:`, like so:

```html
<p class="first-letter:uppercase">hello, how cool is that!</p>
```

Hope it helps! 👍

**Reply by damikun** [reply]
Hi @simonswiss thx for your response i will try this feature.. thx much for your answer...
